operator/pkg/gateway-api/httproute_gamma_reconcile.go

```diff
@@ -0,0 +1,159 @@
+		}
+		scopedLog.WithError(err).Error("Unable to fetch HTTPRoute")
+		return controllerruntime.Fail(err)
+	}
+
+	// Ignore deleted HTTPRoute, this can happen when foregroundDeletion is enabled
+	if original.GetDeletionTimestamp() != nil {
+		return controllerruntime.Success()
+	}
+
+	hr := original.DeepCopy()
+
+	// check if this cert is allowed to be used by this gateway
+	grants := &gatewayv1beta1.ReferenceGrantList{}
+	if err := r.Client.List(ctx, grants); err != nil {
+		return r.handleReconcileErrorWithStatus(ctx, fmt.Errorf("failed to retrieve reference grants: %w", err), original, hr)
+	}
+
+	// input for the validators
+	i := &routechecks.HTTPRouteInput{
+		Ctx:       ctx,
+		Logger:    scopedLog.WithField(logfields.Resource, hr),
+		Client:    r.Client,
+		Grants:    grants,
+		HTTPRoute: hr,
+	}
+
+	// gateway validators
+	for _, parent := range hr.Spec.ParentRefs {
+
+		// set acceptance to okay, this wil be overwritten in checks if needed
+		i.SetParentCondition(parent, metav1.Condition{
+			Type:    string(gatewayv1.RouteConditionAccepted),
+			Status:  metav1.ConditionTrue,
+			Reason:  string(gatewayv1.RouteReasonAccepted),
+			Message: "Accepted HTTPRoute",
+		})
+
+		// set status to okay, this wil be overwritten in checks if needed
+		i.SetAllParentCondition(metav1.Condition{
+			Type:    string(gatewayv1.RouteConditionResolvedRefs),
+			Status:  metav1.ConditionTrue,
+			Reason:  string(gatewayv1.RouteReasonResolvedRefs),
+			Message: "Service reference is valid",
+		})
+
+		// run the actual validators
+		// for _, fn := range []routechecks.CheckGatewayFunc{
+		// 	routechecks.CheckGatewayAllowedForNamespace,
+		// 	routechecks.CheckGatewayRouteKindAllowed,
+		// 	routechecks.CheckGatewayMatchingPorts,
+		// 	routechecks.CheckGatewayMatchingHostnames,
+		// 	routechecks.CheckGatewayMatchingSection,
+		// } {
+		// 	continueCheck, err := fn(i, parent)
+		// 	if err != nil {
+		// 		return r.handleReconcileErrorWithStatus(ctx, fmt.Errorf("failed to apply Gateway check: %w", err), original, hr)
+		// 	}
+
+		// 	if !continueCheck {
+		// 		break
+		// 	}
+		// }
+	}
+
+	for _, fn := range []routechecks.CheckRuleFunc{
+		routechecks.CheckAgainstCrossNamespaceBackendReferences,
+		routechecks.CheckBackend,
+		routechecks.CheckBackendIsExistingService,
+	} {
+		continueCheck, err := fn(i)
+		if err != nil {
+			return r.handleReconcileErrorWithStatus(ctx, fmt.Errorf("failed to apply Backend check: %w", err), original, hr)
+		}
+
+		if !continueCheck {
+			break
+		}
+	}
+
+	if err := r.updateStatus(ctx, original, hr); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to update HTTPRoute status: %w", err)
+	}
+
+	scopedLog.Info("Successfully reconciled HTTPRoute")
+	return controllerruntime.Success()
+}
+
+func (r *gammaHttpRouteReconciler) updateStatus(ctx context.Context, original *gatewayv1.HTTPRoute, new *gatewayv1.HTTPRoute) error {
+	oldStatus := original.Status.DeepCopy()
+	newStatus := new.Status.DeepCopy()
```

Comment:
Do we need to do a deep copy before checking for equalness?

Reply:
We don't do locking of the controller-runtime cache, so it's intended to ensure that we don't have to worry about the HTTPRoute changing during the comparison. The API server will throw away status updates that are out of order or no-ops, so it saves some concurrency management to deepcopy here.

Reply:
If we don't do locking doesn't that also mean that the HTTP Route can change while we are doing the DeepCopy?